### ISSUE TITLE
Output of a model (e.g. logits for the first token) depends on input length

### ISSUE THREAD

**Issue by tomasz-kielbasa** [huggingface/transformers#26462]
### System Info

- `transformers` version: 4.33.3
- Platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.31
- Python version: 3.9.16
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.3.3
- Accelerate version: 0.23.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): 2.9.2 (True)
- Flax version (CPU?/GPU?/TPU?): 0.6.3 (gpu)
- Jax version: 0.4.1
- JaxLib version: 0.4.1
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", device_map='auto')

text = 'This is a'
input_ids = tokenizer(text, return_tensors='pt').input_ids.to('cuda')

for i in range(input_ids.shape[1]):
    full_text_logits = model(input_ids).logits[0][0]
    single_token_logits = model(input_ids[:, :i + 1]).logits[0][0]
    
    error = (full_text_logits - single_token_logits).abs()
    print(error.sum().item(), error.max().item())
```
Output:
```
0.5521924495697021 8.106231689453125e-05
0.10465995222330093 2.002716064453125e-05
0.06035400182008743 8.58306884765625e-06
0.0 0.0
```

### Expected behavior

The above numbers should be zero or closer to zero. A total error of 0.55 looks suspicious. Tried also with a different model (gpt2) with similar results. If this is the desired behavior, is there a workaround to achieve deterministic results?

**Comment by LysandreJik** [comment]
cc @ArthurZucker

**Comment by tomasz-kielbasa** [comment]
It seems this is a PyTorch/CUDA thing.
```python
import torch
from torch import nn
torch.manual_seed(0)

linear = nn.Linear(4, 4).to('cuda')
dummy_input = torch.normal(mean=0, std=1, size=(2, 4)).to('cuda')
full = linear(dummy_input)[0]
single = linear(dummy_input[:1])[0]
full == single
```
This code results in the following output.
```
tensor([ True, False,  True,  True], device='cuda:0')
```
Note that if we run `linear` multiple times on the exact same input, we get the same output, so this problem isn't about determinism, but consistency.
